src/connector/src/parser/dynamodb/json_parser.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use assert_matches::assert_matches;
+use risingwave_common::types::DataType;
+use risingwave_connector_codec::decoder::Access;
+
+use crate::error::ConnectorResult;
+use crate::only_parse_payload;
+use crate::parser::dynamodb::{build_dynamodb_json_accessor_builder, map_rw_type_to_dynamodb_type};
+use crate::parser::{
+    AccessBuilderImpl, ByteStreamSourceParser, ParserFormat, SourceStreamChunkRowWriter,
+    SpecificParserConfig,
+};
+use crate::source::{SourceColumnDesc, SourceContext, SourceContextRef};
+
+const ITEM: &str = "Item";
+
+// An example of a DynamoDB event on S3:
+// {
+//     "Item":{
+//         "customer_name":{
+//             "S":"Bob Lee"
+//         },
+//         "order_id":{
+//             "N":"3"
+//         },
+//         "price":{
+//             "N":"63.06"
+//         },
+//         "order_status":{
+//             "N":"3"
+//         },
+//         "product_id":{
+//             "N":"2060"
+//         },
+//         "order_date":{
+//             "N":"1720037677"
+//         }
+//     }
+// }
+#[derive(Debug)]
+pub struct DynamodbJsonParser {
+    payload_builder: AccessBuilderImpl,
+    pub(crate) rw_columns: Vec<SourceColumnDesc>,
+    source_ctx: SourceContextRef,
+    single_blob_column: String,
+}
+
+impl DynamodbJsonParser {
+    pub async fn new(
+        props: SpecificParserConfig,
+        rw_columns: Vec<SourceColumnDesc>,
+        source_ctx: SourceContextRef,
+    ) -> ConnectorResult<Self> {
+        let (payload_builder, single_blob_column) =
+            build_dynamodb_json_accessor_builder(props.encoding_config).await?;
+        Ok(Self {
+            payload_builder,
+            rw_columns,
+            source_ctx,
+            single_blob_column,
+        })
+    }
+
+    pub async fn parse_inner(
+        &mut self,
+        payload: Vec<u8>,
+        mut writer: SourceStreamChunkRowWriter<'_>,
+    ) -> ConnectorResult<()> {
+        let payload_accessor = self.payload_builder.generate_accessor(payload).await?;
+        writer.do_insert(|column| {
+            if column.name == self.single_blob_column {
+                assert_matches!(column.data_type, DataType::Jsonb);
+                payload_accessor.access(&[ITEM], &column.data_type)
```

Comment:
I'm a little concerned about this: In the result jsonb, user will need to `data->field->S/N` to access the field?

Reply:
Yes, correct.

Reply:
> Yes. It's inevitable anyway, as DynamoDB doesn't not enforce the type of attributes.
data->field can either be an S or N.
->S/N can be viewed as a manual type casting.

I have concerns about value range. Can you help confirm that data from DynamoDB can be correctly represented in JSON?

Reply:
https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/S3DataExport.Output.html?icmpid=docs_dynamodb_help_panel_hp_export_format

Reply:
https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_Record.html

Reply:
> It's inevitable anyway

Why it's "inevitable"? Actually I had thought introducing `FROMAT DYNAMODB` is because we want to treat it specially. (like no need to access type, but parser does it implicitly) But now it seems it's not different from `FORMAT PLAIN` (just syntactic sugar?).

BTW, how does other systems' dynamodb CDC's behavior? Will they also need to access the type field?

I'm not arguing we should do that. Just express confusion.

Reply:
https://www.youtube.com/live/JwDbdsQPWak 
at 27:05
You can see they just keep the how event as a json.

Reply:
As they are not a database, they didn't try to understand the json though, basically they only parsed the top layer into the table. We can also parse the system info (like `eventID` and `awsRegion`) besides the primary keys and main events, if neccessary.

Reply:
> >Numbers can be positive, negative, or zero. Numbers can have up to 38 digits of precision. Exceeding this results in an exception. If you need greater precision than 38 digits, you can use strings.
> 
> Thanks @tabVersion for reminding this.
It indeed has a higher precision compared to our decimals (28 digits). 😢
How did we typically address such cases? cc @xiangjinwu

It is why we did not encode the whole row into a single jsonb column.
JSON has lower precision in its standard. We can do little about it as an application. 

> Why it's "inevitable"? Actually I had thought introducing FROMAT DYNAMODB is because we want to treat it specially. (like no need to access type, but parser does it implicitly) But now it seems it's not different from FORMAT PLAIN (just syntactic sugar?).

Agree, I prefer letting users define the schema themselves. Wrapping it as an individual format is overkill and I don't see a great advantage from it.
We may pivot to support tools generating a `create source` statement from one message instead.

> Define a schema beforehand in RW, and skip records that do not match the schema. I think this goes against dynamodb's nosql design. (Decodable)

But RisingWave serves as a relational database for streaming jobs and analysis. How do users run a streaming job without a clear definition of what the job does? If the user is unsure about a column's content, keeping it as bytes may keep the best compatibility, which we already supported.